base/arrayshow.jl

```diff
@@ -58,9 +58,9 @@ Alignment is reported as a vector of (left,right) tuples, one for each
 column going across the screen.
 """
 function alignment(io::IO, X::AbstractVecOrMat,
-        rows::AbstractVector, cols::AbstractVector,
-        cols_if_complete::Integer, cols_otherwise::Integer, sep::Integer)
-    a = Tuple{Int, Int}[]
+        rows::AbstractVector{T}, cols::AbstractVector{V},
+        cols_if_complete::Integer, cols_otherwise::Integer, sep::Integer) where {T,V}
+    a = Tuple{T, V}[]
```

Comment:
This change is more concerning. I'm a little surprised that we don't `@nospecialize(X)`. Julia ships with 19 precompiled MethodInstances whose backedges include `print_matrix`, and these are pretty expensive to recompile:
```julia
julia> include("/tmp/sv.jl")
2-element SquaresVector:
 1
 4  0.395123 seconds (1.06 M allocations: 58.098 MiB, 6.47% gc time, 99.91% compilation time)
```
where `sv.jl` is taken from https://docs.julialang.org/en/v1/manual/interfaces/#man-interface-array and then adds display:
```julia
struct SquaresVector <: AbstractArray{Int, 1}
    count::Int
end
Base.size(S::SquaresVector) = (S.count,)
Base.IndexStyle(::Type{<:SquaresVector}) = IndexLinear()
Base.getindex(S::SquaresVector, i::Int) = i*i

sv = SquaresVector(2)
@time show(stdout, MIME("text/plain"), sv)
```
0.4s is a *long* time, and we have to pay it if stuff gets invalidated.

Since the `rows` and `cols` are supposed to be lists of indices, I'll ask whether this change is truly needed or whether there might be another way? We'll do it if we must, but try looking at
```julia
code_warntype(Base.alignment, (IO, AbstractVecOrMat, AbstractVector, AbstractVector, Integer, Integer, Integer))
```
and
```julia
code_warntype(Base.print_matrix, (IO, AbstractVecOrMat, AbstractString, AbstractString, AbstractString, AbstractString, AbstractString, AbstractString, Integer, Integer))
```
to see the invalidation risks and how much the type-asserts save us.

Reply:
> Since the rows and cols are supposed to be lists of indices, I'll ask whether this change is truly needed or whether there might be another way? 

It seems what we want to say is that we are iterating over the "range" of a vector. A more descriptive way would be to change it to `rowsA, colsA =  union(axes(X,1)),union(axes(X,2))`. Unfortunately, this won't work since `union(::AbstractRange)` is not specialised:
```julia
julia> union(Base.OneTo(5))
5-element Vector{Int64}:
 1
 2
 3
 4
 5
```
There's a very good argument for adding special overloads for `union(::AbstractRange)`, though making the return type immutable would be a breaking change. And this wouldn't actually improve any of the invalidation issues.